### PR TITLE
Refactor Serializer

### PR DESCRIPTION
**Pull request by rafaelraposospot**
# TL;DR
Refactor serializer and implement missing Kinds

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
_Remove the '*fixes*' keyword if there will be multiple PRs to fix the linked issue_

fixes https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

File: flytekit-examples/src/main/java/org/flyte/examples/SimpleStructTask.java [hunk 1]
```diff
@@ -16,10 +16,15 @@
...
+import org.flyte.api.v1.Struct;
+import org.flyte.flytekit.SdkBindingData;
+import org.flyte.flytekit.SdkRunnableTask;
+import org.flyte.flytekit.SdkTransform;
+import org.flyte.flytekit.jackson.JacksonSdkType;
+
 /** Example Flyte task that takes a name as the input and outputs a simple greeting message. */
-/*
-@AutoService(SdkRunnableTask.class)
+// @AutoService(SdkRunnableTask.class)
```
Comment (andresgomezfrr) [inline flytekit-examples/src/main/java/org/flyte/examples/SimpleStructTask.java:27]: It is fine? Do we want to remove this AutoService?

**Review by pablocasares — APPROVED** [review]
LGTM

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/LiteralSerializer.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+      LiteralType literalType) {
+    this.gen = gen;
+    this.key = key;
+    this.value = value;
+    this.serializerProvider = serializerProvider;
+    this.literalType = literalType;
+  }
+
+  final void serialize() throws IOException {
+    gen.writeFieldName("literal");
```
Comment (andresgomezfrr) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/LiteralSerializer.java:47]: I think that we talked about creating a Constant for `"literal"`, right?

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/BlobSerializer.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+      String key,
+      Literal value,
+      SerializerProvider serializerProvider,
+      LiteralType literalType) {
+    super(gen, key, value, serializerProvider, literalType);
+  }
+
+  @Override
+  void serializeScalar() throws IOException {
+    gen.writeFieldName("scalar");
```
Comment (andresgomezfrr) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/BlobSerializer.java:39]: I think that we talked about creating a Constant for `"scalar"`, right?

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/CollectionSerializer.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    }
+  }
+
+  @Override
+  void serializeLiteral() throws IOException {
+    gen.writeObject(Literal.Kind.COLLECTION);
+    LiteralType elementType = literalType.collectionType();
+    LiteralTypeSerializer.serialize(elementType, gen);
+
+    gen.writeFieldName("value");
```
Comment (andresgomezfrr) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/CollectionSerializer.java:45]: I think that we talked about creating a Constant for `"value"`, right?

Comment (narape) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/CollectionSerializer.java:45]: Yes I will create a new constant for those

File: flytekit-examples/src/main/java/org/flyte/examples/SimpleStructTask.java (hunk 1 above)
Comment (narape) [reply flytekit-examples/src/main/java/org/flyte/examples/SimpleStructTask.java:27]: The struct support is out of the scope for this PR so I commented out the AutoService annotation and some strut specific tests to make the build work.

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/CollectionSerializer.java (hunk 4 above)
Comment (rafaelraposospot) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/CollectionSerializer.java:45]: https://github.com/flyteorg/flytekit-java/pull/167

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/BlobSerializer.java (hunk 3 above)
Comment (rafaelraposospot) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/BlobSerializer.java:39]: https://github.com/flyteorg/flytekit-java/pull/167

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/LiteralSerializer.java (hunk 2 above)
Comment (rafaelraposospot) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/LiteralSerializer.java:47]: https://github.com/flyteorg/flytekit-java/pull/167

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+              case DATETIME:
+                return SdkBindingData.ofDatetime(Instant.parse(tree.get(VALUE).asText()));
+              case FLOAT_VALUE:
+                return SdkBindingData.ofFloat(tree.get(VALUE).doubleValue());
+            }
+            break;
+          case GENERIC:
+          case BLOB:
+            // TODO: We need to implement this
+            throw new RuntimeException("not supported");
```
Comment (narape) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:78]: Give better error message and unify the specific exception, for example generic RuntimeException vs IllegalArgumentException vs UnsupportedOperation Exception

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+      case MAP:
+        return transformMap(tree);
+
+      default:
+        throw new UnsupportedOperationException(
+            String.format("Not supported literal type %s", literalKind.name()));
+    }
+
+    // TODO: Think about it
+    throw new IllegalStateException("");
```
Comment (narape) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:93]: Same here

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+        throw new RuntimeException("not supported");
+      case SIMPLE_TYPE:
+      case MAP_VALUE_TYPE:
+      case COLLECTION_TYPE:
+        List<? extends SdkBindingData<?>> x =
+            streamOf(elements).map(this::transform).collect(toList());
+
+        return SdkBindingData.ofBindingCollection((List<SdkBindingData<T>>) x);
+    }
+    return null; // TODO throw exception
```
Comment (narape) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:113]: here

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T> SdkBindingData<Map<String, T>> transformMap(JsonNode tree) {
```
Comment (narape) [inline flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:117]: follow exception handeling done by previous method

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java (hunk 5 above)
Comment (lupasarin) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:78]: Which type of exception do you suggest that we use?

Comment (narape) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:78]: We chatted and decided on `IllegalArgumentException` because the serialization and deserialization are executed because the user did `JacksonSdkType.of(Input.class)`.
```suggestion
            throw new IllegalArgumentException("Type contains an unsupported scalar: " + scalarKind);
```
assuming that we wrote on line 58 
```java
Scalar.Kind scalarKind = Scalar.Kind.valueOf(tree.get(SCALAR).asText());
switch (scalarKind) {
```

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java (hunk 6 above)
Comment (narape) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:93]: Actually I think we can remove this exception if we make sure that the other case or default ends with returns or throwing exception.
I think that the primitive case should have the default clause

File: flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java (hunk 7 above)
Comment (narape) [reply flytekit-jackson/src/main/java/org/flyte/flytekit/jackson/deserializers/SdkBindingDataDeserializer.java:113]: Add the default clause with the 
```java
        throw new UnsupportedOperationException(
            String.format("Not supported literal type %s", kind.name()));
```

**Comment by narape** [comment]
I manually approved the DCO for Rafael but we will squash all of this when we merge to master
